core/modules.rs

```diff
@@ -385,154 +384,112 @@ impl ModuleLoader for NoopModuleLoader {
 pub type ExtModuleLoaderCb =
   Box<dyn Fn(&ExtensionFileSource) -> Result<ModuleCode, Error>>;
 
-pub struct ExtModuleLoader {
+pub(crate) struct ExtModuleLoader {
   module_loader: Rc<dyn ModuleLoader>,
-  extensions: Rc<RefCell<Vec<Extension>>>,
-  ext_resolution_allowed: RefCell<bool>,
-  used_esm_sources: RefCell<HashMap<String, bool>>,
-  maybe_load_callback: Option<ExtModuleLoaderCb>,
-}
-
-impl Default for ExtModuleLoader {
-  fn default() -> Self {
-    Self {
-      module_loader: Rc::new(NoopModuleLoader),
-      extensions: Default::default(),
-      ext_resolution_allowed: Default::default(),
-      used_esm_sources: RefCell::new(HashMap::default()),
-      maybe_load_callback: None,
-    }
-  }
+  maybe_load_callback: Option<Rc<ExtModuleLoaderCb>>,
+  resolutions: RefCell<HashMap<String, Option<ExtensionFileSource>>>,
+  used_specifiers: RefCell<HashSet<String>>,
 }
 
 impl ExtModuleLoader {
   pub fn new(
-    module_loader: Option<Rc<dyn ModuleLoader>>,
-    extensions: Rc<RefCell<Vec<Extension>>>,
-    maybe_load_callback: Option<ExtModuleLoaderCb>,
+    module_loader: Rc<dyn ModuleLoader>,
+    extensions: &[Extension],
+    module_map: &ModuleMap,
+    maybe_load_callback: Option<Rc<ExtModuleLoaderCb>>,
   ) -> Self {
-    let used_esm_sources: HashMap<String, bool> = extensions
-      .borrow()
-      .iter()
-      .flat_map(|e| e.get_esm_sources())
-      .flatten()
-      .map(|file_source| (file_source.specifier.to_string(), false))
-      .collect();
-
+    let mut resolutions = HashMap::new();
+    resolutions.extend(
+      extensions
+        .iter()
+        .flat_map(|e| e.get_esm_sources())
+        .flatten()
+        .map(|s| (s.specifier.to_string(), Some(s.clone()))),
+    );
+    resolutions.extend(
+      module_map
+        .names()
+        .iter()
+        .map(|n| (n.as_str().to_string(), None)),
```

Comment:
I don't understand this bit - why would we have `ModuleMap` created before `ExtModuleLoader`?

Reply:
It's so `ExtModuleLoader` is made aware of extension modules that are already in the snapshot, since those aren't passed in `Extension::get_esm_sources()`. Maybe that info shouldn't be excluded... we can add `ExtensionFileSourceCode::IncludedInSnapshot` which asserts that it never gets loaded.

Reply:
I'd be +1 on that - the more side wheel to ensure nothing goes wrong the better